### ISSUE TITLE
is_instantiable check is broken by nested types

### ISSUE THREAD

**Issue by arielb1** [rust-lang/rust#27497]
## STR
```Rust
struct Foo<T>(T, Box<Bar<Bar<Foo<T>>>);
struct Bar<T>(T);

fn main() {
    println!("Hello, world!");
}
```

## Results

`Foo` is uninhabitable, but the uninhabitability error doesn't trigger. Removing one of the `Bar`-s does make it trigger. This is rather suprising.

The check was added in the olden days when typeck was much less robust, and can be bypassed very easily by associated types anyway - we may want to remove it.

cc @nikomatsakis

**Comment by arielb1** [comment]
`is_instantiable` is also broken when dealing with empty types - it treats them as instantiable, meaning that `struct Foo(Box<Foo>, Void)` is "valid".

**Comment by eefriedman** [comment]
`is_instantiable` errors are already disabled when the `static_recursion` feature gate is turned on.  I agree that it isn't something we really need to check.

**Comment by nikomatsakis** [comment]
I'm trying to remember what motivated us to add that check in the first place, but I cannot.

**Comment by arielb1** [comment]
@nikomatsakis

The old type-checker was *far* less robust then. Can it be removed?

**Comment by nikomatsakis** [comment]
@arielb1 seems ok to me, but cc @rust-lang/lang

**Comment by pnkfelix** [comment]
(I'm okay with removing the check, without going through the RFC process.)

**Comment by nikomatsakis** [comment]
@arielb1 after discussing in lang subteam meeting, general conclusion was "seems fine to experiment with removing the `is_instantiable` check and see what happens". I agree it was dubious to start and is more dubious now with associated types etc.
